### PR TITLE
DO NOT MERGE: Optimizations on a wave operator

### PR DESCRIPTION
**Pull request by kaushikcfd**
Uses
-  Meshmode: https://github.com/kaushikcfd/meshmode/tree/pytato-array-context-transforms
- Pytato: https://github.com/kaushikcfd/pytato/tree/pytato-array-context-transforms
- Arraycontext: https://github.com/kaushikcfd/arraycontext/tree/pytato-array-context-transforms
- Loopy: https://github.com/inducer/loopy/tree/tighter_write_race_checker

### CODE REVIEW

File: examples/wave/wave-opt-3d.py [hunk 1]
```diff
@@ -0,0 +1,821 @@
...
+                for ary in knl.temporary_variables.values():
+                    if ary.address_space == lp.AddressSpace.GLOBAL:
+                        # global temps would be written once and read once
+                        nfootprint_bytes += (2 * product(ary.shape)
+                                             * ary.dtype.itemsize)
+
+                # }}}
+
+                print("Double-prec. GFlOps:", f64_ops * 1e-9)
+                print("Footprint GBs:",  nfootprint_bytes * 1e-9)
```
Comment (kaushikcfd) [inline examples/wave/wave-opt-3d.py:474]: @inducer: Could you take a look at the way we are gathering these metrics. Any issues?

File: grudge/metadata.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from pytools.tag import UniqueTag, tag_dataclass
+from .dof_desc import DOFDesc
+
+
+@tag_dataclass
+class DiscretizationElementAxisTag(UniqueTag):
+    """
+    Tagged to an array's axis travesing elements in a discretization.
```
Comment (inducer) [inline grudge/metadata.py:8]: - typo: traversing
- traversing -> "ranging across"? "enumerating"? "representing element indices"?
- "in a discretization" -> "in the discretization identified by :attr:`dd`"

File: grudge/metadata.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    :attr dd: One of the domain discretization tags defined in
+        :mod:`grudge.dof_desc`.
+    """
+    dd: DOFDesc
+
+
+@tag_dataclass
+class DiscretizationDOFAxisTag(UniqueTag):
+    """
+    Tagged to an array's axis travesing an DOFs in a discretization.
```
Comment (inducer) [inline grudge/metadata.py:19]: typo: traversing

File: grudge/metadata.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from pytools.tag import UniqueTag, tag_dataclass
+from .dof_desc import DOFDesc
+
+
+@tag_dataclass
+class DiscretizationElementAxisTag(UniqueTag):
```
Comment (inducer) [inline grudge/metadata.py:6]:
```
class AxisTypeTag(UniqueTag):
    ...

class DiscretizationElementAxisTag(AxisTypeTag):
    ...

class DiscretizationDOFAxisTag(AxisTypeTag):
    ...

File: grudge/metadata.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    Tagged to an array's axis travesing an DOFs in a discretization.
+
+    :attr dd: One of the domain discretization tags defined in
+        :mod:`grudge.dof_desc`.
+    """
+    dd: DOFDesc
+
+
+@tag_dataclass
+class DicretizationElementProjector(UniqueTag):
```
Comment (inducer) [inline grudge/metadata.py:28]: - typo: Discretization

File: grudge/metadata.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    """
+    Tagged to an array used an indirection map from one discretization's
+    elements to another discretrization's elements.
+    """
+    from_dd: DOFDesc
+    to_dd: DOFDesc
+
+
+@tag_dataclass
+class DicretizationDOFProjector(UniqueTag):
```
Comment (inducer) [inline grudge/metadata.py:38]: - typo: Discretization

File: grudge/metadata.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+@tag_dataclass
+class DiscretizationElementAxisTag(UniqueTag):
+    """
+    Tagged to an array's axis travesing elements in a discretization.
+
+    :attr dd: One of the domain discretization tags defined in
+        :mod:`grudge.dof_desc`.
+    """
+    dd: DOFDesc
```
Comment (inducer) [inline grudge/metadata.py:13]: Add:

- `group_index`
- `grp.discretization_key`

**Comment by inducer** [comment]
This is quite possibly old news at this point. Ready to close?

**Comment by kaushikcfd** [comment]
>  This is quite possibly old news at this point. Ready to close?

Yep!

**Pull request closed without merging**